app/src/main/java/app/grapheneos/camera/ui/activities/MainActivity.kt

```diff
@@ -1058,16 +1058,35 @@ open class MainActivity : AppCompatActivity(),
     private lateinit var dialog: Dialog
 
     open fun bytesToHex(bytes: ByteArray): String {
-        val hexChars = CharArray(bytes.size)
-        for (i in 0 until bytes.indices.last - 1 step 3) {
-            val v: Int = bytes[i].toInt() and 0xFF
-            hexChars[i] = hexArray[v ushr 4]
-            hexChars[i + 1] = hexArray[v and 0x0F]
-            hexChars[i + 2] = ' '
+        // Add a space between every byte
+        // Double space every 4 bytes (unless end or newline)
+        // Add a newline every 8 bytes (unless end)
+
+        var outLen = bytes.size * 3 // 2 hex digits + 1 space/newline per byte
+        outLen += bytes.size / 4 // Double spaces
+        outLen -= bytes.size / 8 // No double space when newline
+        outLen -= 1 // No trailing space/newline
+        if (bytes.size % 4 == 0 && bytes.size % 8 != 0)
+            outLen -= 1 // No trailing double space
+
+        val hexChars = CharArray(outLen)
```

Comment:
for an empty `ByteArray`, `outLen` will be `-1`, which will cause exception when `CharArray` is initialised (might crash the app, if not handled). early returning an empty string (if the size of `ByteArray` is zero) can mitigate this